TODO1/tracing-or-debugging-vue-js-reactivity-the-computed-tree.md

```diff
@@ -2,143 +2,143 @@
+但我们看看派生的数据，它的情况有点复杂。首先，计算属性的值是被缓存起来的，以便在它算出来之后就一直返回原值，只有当它的缓存失效才会被重新计算，换句话说，只在其依赖的数据发生改变时它们才会重新求值。
 
-Let’s look at part of the [JSFiddle](https://jsfiddle.net/mikeapr4/eqLy1ac3/) example. The `currentUserId` state property is used in the `currentUser` getter, then in the `validCurrentUser` computed property, which in turn is part of a `v-if` expression in the render of the root component. This is a nice chain to look at.
+我们再来看看[之前的例子](https://jsfiddle.net/mikeapr4/eqLy1ac3/)。`currentUserId` 状态被 `currentUser` 这个 getter 引用了，然后在 `validCurrentUser` 计算属性引用了 `currentUser`，`validCurrentUser` 又是根组件 render 函数的 `v-if` 表达式的一部分。这条引用链看起来不错。
 
-In practice, memoization is handled by a configuration option to the Watcher. When we use Watchers from components the [API documentation](https://vuejs.org/v2/api/#vm-watch) describes 2 possible options (`deep`, `immediate`) but there are more undocumented options, I’m not suggesting you use them, but worth understanding them. One option is `lazy`, which means that the Watcher will maintain a flag (`dirty`) that will be true if the reactive data has changed but the Watcher has not yet re-run, i.e. the cache is stale.
+实际上，响应数据的存储是通过一个 Watcher 的配置选项来处理的。当我们使用组件中的 Watcher 时，[API 文档](https://vuejs.org/v2/api/#vm-watch)中介绍了两个可选选项（`deep`，`immediate`），但其实还有一些没被文档记录的选项，我并不推介你使用这些没被记录的选项，但理解他们却很有益处。其中一个选项是 `lazy`，配置它之后 Watcher 将会维护一个 `dirty` 标志，如果依赖的响应数据已经更改但这个 Watcher 还未运行时它将为 true，也就是说，此时缓存已过时。
 
-So what happens in our example when `currentUserId` is changed to, say, 3. Any lazy Watchers will be flagged as dirty straight away, but not run. Both `currentUser` and `validCurrentUser` are lazy watchers of this state property. The root render function will also depend on this state property, it will fire on next tick. When it runs it will call `validCurrentUser`, now flagged dirty, which will re-run its getter function, in turn calling `currentUser` which also will update. And that’s it, the component has re-rendered correctly and caches have been refreshed.
-> # Hold up I hear you say, why would all 3 Watchers depend on the state property?
+在我们的例子中，如果 `currentUserId` 被改成 3。任何依赖于它且被设置了 `lazy` 的 Watcher 都会被标记为 dirty，但 Watcher 并没有运行。`currentUser` 和 `validCurrentUser` 都是这个状态的 lazy Watcher。根渲染函数同样会依赖于这个状态，渲染将在下一个 tick 时被触发。当渲染函数执行时，将会访问已经被标记为 dirty 的 `validCurrentUser`，它将重新运行它的 getter 函数，进而访问同样需要更新的 `currentUser`。至此，这个组件将会被正确重渲染，并且相关缓存将被更新。
+> # 等等，我似乎听见你在问，为什么所有 3 个 Watcher 都是依赖于这个状态的呢？
 
-Don’t they depend on each other? Well, a feature of computed watchers is that their values aren’t reactive, but rather when a computed getter is called, all of its own dependencies are passed to the calling Watcher, if there is one. This flattening of reactivity chains is better for performance not to mention being a simpler solution.
+难道他们不是相互依赖的么？计算属性 watcher 有一个特性就是不仅它自身的值是响应式的，而且当计算属性的 getter 被调用时，如果当前有 Wathcer 在读取这个计算属性的话（即 `Dep.target` 中有值--译者），所有这个计算属性的依赖也将会被这个 Wathcer 收集起来。这种依赖收集关系链的扁平化对性能表现更优，而且也是个比较简单的解决方案。
```

Comment:
【计算属性 watcher ............. 被这个 Watcher 收集起来】 => 【计算属性 Watcher 的一个特性就是它们自身的值不是被动的， 而是当调用一个计算的 getter 时，如果它存在依赖关系，它会把自己所有的依赖关系都传递给调用的 Watcher】

Reply:
这里有点难翻，`their values aren’t reactive` 这个reactive是被动的还是响应式的呢，我觉得可能想表达的是，它不仅自己是响应式的，还做了其他的事，事情就是后面 `rather than` 介绍的
`not ... rather than` 是个词组，不仅...而且更是

Reply:
我感觉像是表达转折 😂 ，看一下另一位校对 @hanxiansen 的意见？我也举得这个地方有待商讨， 如果确定一下 reactive 想表达的意思也就比较好继续了。我觉得你说的这种情况也是有道理的。

Reply:
@hanxiansen  这里你怎么看？